### PR TITLE
Add more specific error thowring base on PR 1918

### PR DESCRIPTION
**Pull request by marcofognog**
Closes https://github.com/ranaroussi/yfinance/issues/270 and https://github.com/ranaroussi/yfinance/issues/1925

## Purpose

It is useful to have the lib to throw specific errors for each failure, instead of the generic ```Exception```.

With the error types being thrown, clients can chose to handle each failure scenario in different ways.

Ex:

```python
    try:
        hist = ticker.history(period=period, interval=interval, raise_errors=True)
    except YFPricesMissingError as ex: 
            flash('The provided ticker symbol was not found, perhaps you mispelled it?')
    except YFInvalidPeriodError
            flash('Please provide a valid period')
            return redirect('/')
 ```

## Backwards compatility

Because currently the exception being raise is the type `Exception`, we can safely assume no one is rescuing any of those new types. If any one is rescuing `Exception` code will continue to work in the same way.

## Notes

This PR is basically the same as https://github.com/ranaroussi/yfinance/pull/1918 except I rebased to current `dev` and implemented changes suggested on the original PR. 

I also tested it with my client app and it works as expected.

### CODE REVIEW

**Comment by paulmcq** [comment]
Typo:
'''class YFTickerMissingError(YFException):
    def __init__(self, ticker, rationale):
        super().__init__(f"${ticker}: possibly delisted; {rationale}")
        self.rationale = rationale
        self.ticker = **ticker**  ## _rationale_ 
'''

**Comment by marcofognog** [comment]
> Typo: '''class YFTickerMissingError(YFException): def **init**(self, ticker, rationale): super().**init**(f"${ticker}: possibly delisted; {rationale}") self.rationale = rationale self.ticker = **ticker** ## _rationale_ '''

I'm sorry, I still can't see the typo, could you be more specific?

**Comment by paulmcq** [comment]
At line 25 of yfinance/exceptions.py. the new code says

self.ticker = *rationale*

but surely that should be

self.ticker = *ticker*

because line 24 is

self.rationale = rationale


Sorry I was not more explicit.

**Comment by marcofognog** [comment]
> At line 25 of yfinance/exceptions.py. the new code says self.ticker = rationale but surely that should be self.ticker = ticker because line 24 is self.rationale = rationale Sorry I was not more explicit.

Good catch! Thank you very much.

**Comment by marcofognog** [comment]
FYI @elibroftw

**Review by ValueRaider — COMMENTED** [review]
Final point: can you git-squash some commits to shorten history, at least your last few commits? Helps this https://github.com/ranaroussi/yfinance/network

File: Pipfile
[the remark this answers is not in the document]
Comment (marcofognog) [reply Pipfile:1]: Oops, this last commit should not be here, sorry for the noise.

**Comment by marcofognog** [comment]
> Final point: can you git-squash some commits to shorten history, at least your last few commits? Helps this https://github.com/ranaroussi/yfinance/network

Sure. I'm squashing only mine to not loose the author on the previous commits.

**Review by ValueRaider — APPROVED** [review]

**Comment by paulmcq** [comment]
Line 25 of exceptions.py is still wrong:
https://github.com/ranaroussi/yfinance/pull/1928#issuecomment-2105027085

**Comment by ValueRaider** [comment]
https://github.com/ranaroussi/yfinance/blob/049337327ea5c4fe97039ada564d227e7e00835f/yfinance/exceptions.py#L25

Looks fine to me

**Comment by paulmcq** [comment]
You are correct. I was apparently looking at the wrong version. Please excuse the spam.

**Pull request merged**